src/Signal-Estimator-GUI/mainwindow.cpp

```diff
@@ -0,0 +1,304 @@
+    std::sort(in_current.begin(),in_current.end(),QPFcompare());
+    QVector<QPointF> out_current = out_data.getCurrentPoints();
+    std::sort(out_current.begin(),out_current.end(),QPFcompare());
+    //this->ui->ErrorLabel_2->setText(QString("updating graphs"));
+
+    this->curve1->setSamples(in_current);
+    this->curve2->setSamples(out_current);
+
+   // this->ui->InputSig->updateAxes();
+    //this->ui->InputSig->replot();
+
+    this->ui->OutputSig->updateAxes();
+    this->ui->OutputSig->replot();
+    QApplication::processEvents();
+
+    if (this->update_plots == true){ // restart timer for graphing
+        this->timer->start();
+    }
+}
+
+
+
+void MainWindow::read_graph_data(){
+     std::string buffer;
+
+     char* fullstr;
+     QByteArray arr;
+
+     this->proc->setReadChannel(QProcess::StandardOutput);
+     arr = this->proc->readLine(); // read line from proc
+     buffer = arr.toStdString(); // convert to std::string
+     //this->ui->ErrorLabel_2->setText(QString(arr));
+
+     fullstr = &buffer[0]; // convert to char* for strtok parsing;
+
+     parseLine(fullstr);
+}
+
+QStringList MainWindow::set_up_program(){
+    // This program calls the separate signal_estimator cmd line program to get the graph data
+    QStringList list; // to convert from QString to normal string
+    QString temp;
+    std::string buffer;
+
+    list.append(" -m");
+    list.append(this->ui->Modes->currentText());
+
+    char* c;
+
+    temp = this->ui->OutputDevices->currentText();
+    // there might be a problem when there is too much arguments.
+    // the arguments start to corrupt when there is too many of them
+    // so for a lot of options we only add them as arguments when they are not default value
+    list.append("-o");
+    if (temp.toStdString() == "default")
+        list.append(temp);
+    else {
+        c = (std::strstr(temp.toStdString().c_str() , "card ") + 5); // get card number
+        buffer = "hw:";
+        buffer.push_back(*c); // hw:X
+        c = (std::strstr(temp.toStdString().c_str() , " device ") + 8); // get device
+        buffer.append(","); // hw:X,
+        buffer.push_back(*c); // hw:X,Y
+        list.append(QString::fromStdString(buffer));
+    }
+
+    temp = this->ui->InputDevices->currentText();
+    list.append("-i");
+    if (temp.toStdString() == "default")
+        list.append(temp);
+    else {
+        c = (std::strstr(temp.toStdString().c_str() , "card ") + 5); // get card number
+        buffer = "hw:";
+        buffer.push_back(*c); // hw:X
+        c = (std::strstr(temp.toStdString().c_str() , " device ") + 8); // get device
+        buffer.append(","); // hw:X,
+        buffer.push_back(*c); // hw:X,Y
+        list.append(QString::fromStdString(buffer));
+    }
+
+    QString t;
+    if ((t = this->ui->SampleRate->cleanText()) != QString::fromStdString("48000")){ // if it is not default value
+        list.append("-r");
+        list.append((t));
+    }
```

Comment:
Why do we need to hardcode all default values here? This code will silently break if the default changes and we forget to update it here. Can't we just always specify all options, even if they are default?